Improvement: Overhaul arrays and objects DQL docs

mechanomi opened the pull request:
## Summary of the changes / Why this improves CrateDB

parent issue: https://github.com/crate/crate/issues/10110

I am leaving the commit history on this PR as-is in case it helps with the review. once I have feedback, I will update the branch and squash the commits

not sure how the commits should be squashed before merging. two options come to mind:

- squash the whole thing into one commit
- squash the changes to `selects.html` (along with other RST files and test data) as an update to the docs, and then squash the `DepricationWarning` fix (and any other related changes that might be necessary) into a separate commit as an update to the build system

maybe there are other ways of doing this? open to suggestions!

all of the changes on this PR result directly from my attempts to improve the arrays and objects SQL examples (and explanatory text) in `selects.html`

the changes on this PR do the following:

- fix a `DepricationWarning` output by `test_docs.py` (see https://github.com/crate/crate/pull/10866/commits/6bda89c20d7d6a1d1d6002a3a4336041b7d6a028, https://github.com/crate/crate/pull/10866/commits/a905e065d7b37d21b1191732d054de4ea7119894, https://github.com/crate/crate/pull/10866/commits/1cd25a2f7575d6f2ef7f58f1f9fdf5a72416e13c) -- encountered while trying to fix the `itests` Gradle task

- renames the `locations` column `race` to `inhabitants` (rationale here: https://github.com/crate/crate/issues/10110#issuecomment-744733672) -- appears in the SQL examples. fixing this necessitated small changes in other files to reflect the rename

- fixes the test data's handling of ID values by removing the double quotes (see [locations.json](https://github.com/crate/crate/pull/10866/files#diff-37950ed9a2670b859a8c005330059afcfa181e2b49ae0eb3d74fc93f85d8e71e), [locations_import_summary1.json](https://github.com/crate/crate/pull/10866/files#diff-13af26bb80f3d64b2cf36a2ffb8730ab02e203d7fd3c5958c909397de81fa240), [locations_import_summary2.json](https://github.com/crate/crate/pull/10866/files#diff-2e587965b6728355b91f7a63d4b56b8d130fc3226999b60272c046824ad55847)). inserting values as strings works, but it results in string sorting (as opposed to numerical sorting) when used in the examples. the previous sort order results in examples that list "1" followed by "10" (`sort by id limit 2`). with this change, the same example lists "1" followed by "2" (which I think is easier to understand, and is more in line with other test data)

- alters the `interests` of two places in the `locations` test data (see [locations.json](https://github.com/crate/crate/pull/10866/files#diff-37950ed9a2670b859a8c005330059afcfa181e2b49ae0eb3d74fc93f85d8e71e)) and adds a `landmarks` column (see [test_docs.py](https://github.com/crate/crate/pull/10866/files#diff-c486a40c01dda609a8a20877e83678604e8dfac6346fb38871fe1b7c570308ef)) so that example queries in the `ANY` section more clearly demonstrate what is being shown

- updates a reference in [table-functions.rst](https://github.com/crate/crate/pull/10866/files#diff-06540fe57d74e90b4f4c10c33e713f5e947c29378ed5071c71432f236bd08e23) to reflect some slight re-working of the sections in `selects.html` (plus zaps some trailing whitespaces -- an automatic thing my editor does. this is good style, so I left it in. but can remove, here and elsewhere, if requested)

- updates any RST files that need updating to reflect the new `locations` test data ([information-schema.rst](https://github.com/crate/crate/pull/10866/files#diff-081d6704593f82a2868419bd2cdf8e758653410c04b4ef6585d88dac6d6e4465), [http.rst](https://github.com/crate/crate/pull/10866/files#diff-7f8aaf9247abd2826b4be0243a604acaabcbdc5367263b0db475f698851817d5), [value-expressions.rst](https://github.com/crate/crate/pull/10866/files#diff-b6d9c695416d2cd5b3da646fdc5b3e62e6565e78944f38a68103604086a75ead))

- for [selects.html](https://github.com/crate/crate/pull/10866/files#diff-cf280b3c8351b1cbc17f43814ad519b3bd84f689567c804270874604fe0aba2c)

  - rework the opening info (see line 7 onwards)
  - updates examples to match the test data in the introduction section (line 27)
  - rewraps some paras to fit 79 chars (line 67 and elsewhere)
  - changes the example output integer column formatting to right-align the value (to match the actual output) (line 191 and elsewhere)
  - moves link lists to the bottom of the file per the [RST style guide](https://github.com/crate/crate-docs/blob/main/style/rst.rst#links) (line 306)
  - removes a redundant example (line 494)
  - re-orders the section so that concepts are introduced gradually and each explanation builds upon the previous one. the new order: *Arrays*, *Objects*, *Arrays within objects*, *Object within arrays*, *ANY (array)*, and *Negating ANY* -- NOTE: I adopted the phrases "arrays within objects" and "objects within arrays" because, in English, "object arrays" is ambiguous and could mean "arrays that belong to an object" or "arrays of objects". this ambiguity significantly adds to the confusion of the existing text
  - in the sections that cover arrays and objects, I reworked the examples so the reader is first shown the data and then shown how to operate on it in successively complex ways. this directly addresses @proddata's observation on https://github.com/crate/crate/issues/10110 that "it is not clear if [the] information [being operated on] is an object or array"
  - moved some info to admonition blocks and added new admonitions to clarify various points (as necessary)
  - changed the example queries to insert more places from the Hitchhikers Guide to the Galaxy universe (in keeping with the theme already established by the test data). as it was, inserting European placenames into this data set produced results that mix two different themes (which adds a small amount of confusion to an already confusing topic)
  - did my best to disambiguate what is happening when the examples query nested arrays and objects. I went through a bunch of iterations on this. it is difficult to explain without making someone dizzy. I think the approach I settled on works though. basically: be deliberate and verbose (e.g., the use of "child" and "parent")
  - added new admonitions (line 697) explaining that the child array index `n` must come before the parent object `property` name (e.g., `my_column[n]['property']`). this is so unintuitive that I was tempted to add a title to this admonition saying something like "SQL syntax Gotcha", but decided against it. I ended up trying to find out if it has been documented before in some sort of PostgreSQL "gotchas" list (e.g., https://sql-info.de/postgresql/postgres-gotchas.html). I couldn't find anything. I am also not clear why we do it this way. is it for compatibility with PostgreSQL? does PostgreSQL not support the inverse syntax? (I do not explain why we do it this way in the text. but if it is useful to do so, I can add that)
  - added notes about the fact that the `my_column[n]['property']` syntax is polymorphic. at least, this is my understanding of it based on my experimentation. it is not mentioned in our docs, and I was unable to find anything useful on the web that would help me explain this. I am not sure polymorphic is the technically correct term to use here, but it seems to fit
  - added a note about the `my_column['property']::text[][1]` syntax mentioned by @jordi in https://github.com/crate/crate/issues/10110#issuecomment-647473814 -- this is a candidate for removal, imo. not sure if it's necessary or useful to add this or whether it just adds more complexity to one of the most complex sections in this document
  - reworked the examples in *ANY* and *Negating ANY* so that the examples better illustrate the purpose of the syntax being demonstrated

## Checklist

 - [ ] Added an entry in `CHANGES.txt` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

mechanomi commented:
this is (finally) passing on Jenkins but it is failing on my local machine and I cannot figure out why

I figured my incidental fix for the `DepricationWarning` output was the culprit so I backed out my changes to `test_docs.py`. still does not work

the two errors (that I have found so far -- the tests are randomized) are:

```
Failed example:
    pretty_print(sh("""curl -isSX PUT '127.0.0.1:4200/_blobs/myblobs/4a756ca07e9487f482465a99e8286abc86ba4dc7' -d 'contents'
    """).stdout.decode("utf-8"))
Expected:
    HTTP/1.1 201 Created
    content-length: 0
Got:
    HTTP/1.1 404 Not Found
    content-length: 0
    
    <BLANKLINE>
```

(a lot of others like this in the same file)

and:

```
File "/Users/nomi/GitHub/crate/crate/blackbox/docs/interfaces/http.rst", line 29, in http.rst
Failed example:
    pretty_print(sh(r"""curl -sS -H 'Content-Type: application/json' \
    -X POST '127.0.0.1:4200/_sql' \
    -d '{"stmt":"select name, position from locations order by id limit 2"}'
    """).stdout.decode("utf-8"))
Expected:
    {
      "cols": [
        "name",
        "position"
      ],
      "rows": [
        [
          "North West Ripple",
          1
        ],
        [
          "Outer Eastern Rim",
          2
        ]
      ],
      "rowcount": 2,
      "duration": ...
    }
Got:
    {
      "error": {
        "message": "RelationUnknown[Relation 'locations' unknown]",
        "code": 4041
      }
    }
```

(also many others in the same file)

judging by the nature of these failures, the test data is not being created properly, or else the tests themselves are being executed against a running CrateDB node that does not have the test data

I do not have any other instances of CrateDB running locally, so I think I can rule out the second explanation

both types of error fail when executing a shell command. that's why I thought it was my change that caused this. but the error persists even with the change reverted

running `git clean -xdff $ ./gradlew` and `rm -rf blackbox/.venv` do not fix the issue

mechanomi commented:
I tried again on master as-is and the tests still fail in the same way. I figure this points to a local setup issue. I realized I was running Python 3.8 and that this may not be supported yet. I wiped my Python 3.8 install (to be absolutely sure it was gone) and re-installed Python 3.7. the issue persists. I am stumped

mechanomi commented:
@seut I have updated the description to reflect the changes made. adding you for review specifically because some of the information I have added here is *based on my understanding of CrateDB syntax based on experimentation alone*. I am not 100% sure the way I have explained it is correct as these peculiarities of the syntax are not mentioned anywhere else in the docs (that I can find). for this reason, I would like some close technical scrutiny

justien commented on docs/general/information-schema.rst:
```diff
@@ -288,24 +288,25 @@ table::
-    SELECT 18 rows in set (... sec)
+    SELECT 19 rows in set (... sec)
 
 You can even query this tables' own columns (attention: this might lead to
```
lint: should be "table's"

mechanomi replied:
fixed in https://github.com/crate/crate/pull/10866/commits/3551afc5a97b5b50159d5257da41be7d9b075aa6

seut commented:
@mechanomi In general I would really prefer that each commit just targets a dedicated topic and includes a verbose description of what it is changing. This would have also eliminated the need to describe your commits inside the PR description.
Additionally it helps a lot with reviewing, specially on PR's which lot of changes.
Having commits just titled with "Improve X..." without any description are not a good pattern...
By following this rule, questions on "what should I squash?" won't be necessary anymore.

seut commented on docs/general/dql/selects.rst:
```diff
@@ -5,40 +5,46 @@
-    +----+-------------------+--------------+--------+----------+---------...-+-------...-+-------------+
-    SELECT 1 row in set (... sec)
+    cr> select * from locations order by id limit 2;
+    +----+-------------------+--------------+--------+----------+--------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+---------------------------------------------------------------------------------------+-----------+
```
Can we keep truncating the result table? I think without, readability gets worse.

seut commented on docs/general/dql/selects.rst:
```diff
@@ -59,8 +65,7 @@ Aliases can be used to change the output name of the columns::
 based upon. Can be a single table, many tables, a view, a :ref:`JOIN <sql_joins>`
-or another ``SELECT`` statement. See :ref:`SELECT Reference
-<sql_reference_select>`
+or another ``SELECT`` statement.
```
```suggestion
or another ref:`SELECT <sql_reference_select>` statement.
```

mechanomi replied:
fixed in https://github.com/crate/crate/pull/10866/commits/3be45f0cd09095a71f55099ba3aeb51b6fd3b48e

seut commented on docs/general/dql/selects.rst:
```diff
@@ -511,274 +502,461 @@ does always return ``NULL`` when comparing ``NULL``.
 
-``ANY (array)``
-===============
+Arrays
```
The sections before were about comparison operators, how does a general `Arrays` section fit here?

mechanomi replied:
good point. I moved it because these sections concern operations on arrays, so I figured they should come after that info. but you're right that this info should be a part of the general comparison operators

to solve both issues, I moved the sections up (as suggested) and added a ref to the opening paras of both sections that links to the arrays section so people can jump there if they need to

fix in https://github.com/crate/crate/pull/10866/commits/4ce0d3dbe7933a5e54466abaab23f806bd6da9ee

seut commented on docs/general/dql/selects.rst:
```diff
@@ -511,274 +502,461 @@ does always return ``NULL`` when comparing ``NULL``.
 
-One important thing to notice when using ANY is that negating the ANY operator
-does not behave as negating normal comparison operators.
+   You can nest arrays within arrays, but you can only have one subscript value
```
No, you can **not** nest arrays with in arrays. Defining a columns as `array(array(text))` is not supported.

seut commented on docs/general/dql/selects.rst:
```diff
@@ -511,274 +502,461 @@ does always return ``NULL`` when comparing ``NULL``.
-Inserting objects::
+.. NOTE::
+
+   You can nest arrays and objects however you like, but you can only have one
```
See my comment above, arrays of arrays are not supported.

seut commented on docs/general/dql/selects.rst:
```diff
@@ -511,274 +502,461 @@ does always return ``NULL`` when comparing ``NULL``.
+   You can nest arrays and objects however you like, but you can only have one
+   subscript value per expression. The following won't work:
+
+   ``select my_column[1]['property'][2] from my_table;``
```
So how would if work then? maybe suggest `my_column[1]['property']::text[][1]` ?

seut commented on docs/general/dql/selects.rst:
```diff
@@ -511,274 +502,461 @@ does always return ``NULL`` when comparing ``NULL``.
-    | [{"evolution_level": 4, "population": 12}, {"evolu...": 42, "popul...": 42}] |
-    | [{"evolution_level": 2, "population": 5673745846}]                           |
-    +---------------------------------------------------...------------------------+
+   You can nest arrays and objects however you like, but you can only have one
```
No, see comment above.

seut commented on docs/general/dql/selects.rst:
```diff
@@ -511,274 +502,461 @@ does always return ``NULL`` when comparing ``NULL``.
+   You can nest arrays and objects however you like, but you can only have one
+   subscript value per expression. The following won't work:
+
+   ``select my_column[1]['property'][2] from my_table;``
```
Also see my comment above.
The whole `Object inside arrays` section looks like the repeated version of `Arrays inside objects`. Does this really makes sense like this? Couldn't we fold this together or just highlight important differences and otherwise link to the previous explanations?

seut commented on docs/general/dql/selects.rst:
```diff
@@ -511,274 +502,461 @@ does always return ``NULL`` when comparing ``NULL``.
+.. _sql_dql_any_array:
+
+``ANY (array)``
+===============
```
I'd suggest to move the general array/object handling description outside of the comparison operator list.
Or we could change the header level and move array comparison operator beneath general array and object description.

seut reviewed (commented):
Left some suggestions and corrections. Nice work 👍

mechanomi commented:
@seut I have resolved all of your comments. took more work than I anticipated (lots of revisions and reworking) but I think the result is much better (flows better, more clarity, precision, and so on). thank you!!

I squashed everything into one commit and added a detailed commit message, per your request

seut reviewed (approved):
Lgtm, thanks!
Btw. please don't forget to remove the merge commit before merging it to master.
